lang/cs/Org.Apache.REEF.Network/Group/Driver/Impl/ContextManager.cs

```diff
@@ -0,0 +1,148 @@
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Org.Apache.REEF.Driver.Context;
+using Org.Apache.REEF.Utilities.Diagnostics;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Network.Group.Driver.Impl
+{
+    /// <summary>
+    /// This class is to manage the active contexts in group communications
+    /// </summary>
+    public class ContextManager
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(ContextManager));
+        private readonly ReaderWriterLockSlim _contextLock = new ReaderWriterLockSlim();
+        private readonly int _totalNumberOfContexts;
+        private readonly IDictionary<string, IActiveContext> _activeContexts;
+
+        public ContextManager(int numberOfContexts)
+        {
+            _totalNumberOfContexts = numberOfContexts;
+            _activeContexts = new Dictionary<string, IActiveContext>();
+        }
+
+        /// <summary>
+        /// Check if the number of active context has reached to the total number
+        /// </summary>
+        /// <returns></returns>
+        public bool AllContextReceived()
+        {
+            lock (_contextLock)
+            {
+                if (_activeContexts.Count < _totalNumberOfContexts)
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Get number of active context in the collection
+        /// </summary>
+        public int Count
+        {
+            get { return _activeContexts.Count; }
+        }
+
+        /// <summary>
+        /// Get all current active context in the collection
+        /// </summary>
+        public ICollection<IActiveContext> ActiveContexts
+        {
+            get
+            {
+                return _activeContexts.Values;
+            }
+        }
+
+        /// <summary>
+        /// Add an active context to the collection
+        /// </summary>
+        /// <param name="context"></param>
+        public void AddContext(IActiveContext context)
+        {
+            _contextLock.EnterWriteLock();
+            try
+            {
+                if (_activeContexts.ContainsKey(context.Id))
+                {
+                    Exceptions.Throw(new Exception("The context is already in the collection" + context.Id), Logger);
+                }
+
+                _activeContexts.Add(context.Id, context);
+            }
+            finally
+            {
+                _contextLock.ExitWriteLock();
+            }
+        }
+
+        /// <summary>
+        /// Remove an active context from the collection 
+        /// </summary>
+        /// <param name="contextId"></param>
+        public void RemoveContext(string contextId)
+        {
+            _contextLock.EnterWriteLock();
+            try
+            {
+                if (_activeContexts.ContainsKey(contextId))
```

Comment:
Little bit hesitate. For strict checking and less tolerant, yes, we should throw exception. I will add it for now and see if it would happen in testing.